lua/cfc_time/server/time_tracking.lua

```diff
@@ -132,28 +135,41 @@ function ctime:initPlayer( ply )
             initialTime.seconds = seconds
         end
 
-        hook.Run( "CFC_Time_PlayerInitialTime", ply, isFirstVisit, initialTime )
+        hook.Run( "CFC_Time_PlayerInitialTime", ply, isFirstVisit, initialTime, lastSession )
         sessionTotalTime = sessionTotalTime + initialTime.seconds
 
         ctime.totalTimes[steamID] = sessionTotalTime
         ctime:broadcastPlayerTime( ply, sessionTotalTime, now, 0 )
     end
 
-    storage:PlayerInit( ply, now, function( data )
-        if not IsValid( ply ) then return end
+    local existingSession = self.sessions[steamID]
+    if existingSession then
+        ErrorNoHaltWithStack( "[CFCTime] Player loaded in, but already had a session? " .. steamID )
+        self:untrackPlayer( steamID )
+    end
 
-        local isFirstVisit = data.isFirstVisit
+    storage:PlayerInit( ply, now, function( data )
+        local lastSession = data.lastSession
         local sessionID = data.sessionID
 
+        local session = { joined = now }
         steamIDToPly[steamID] = ply
+        ctime.sessions[steamID] = session
         ctime.sessionIDs[steamID] = sessionID
-        ctime.sessions[steamID] = { joined = now }
 
-        if isFirstVisit then return setupPly( 0, true ) end
+        if not IsValid( ply ) then
+            session.departed = now
+            logger:warn( "Player left before their session was created? Marking them as departed", steamID )
+            return
+        end
```

Comment:
Main fix:

If the player isn't valid by the time their session is initialized in the database, we mark them as departed and let the interval take care of saving the 0-second session to the DB and then untracking them like normal.

Returning early here, or in the next query actually creates more opportunities for dangling sessions.